packages/flutter/lib/src/gestures/converter.dart

```diff
@@ -70,54 +73,19 @@ class PointerEventConverter {
       final Duration timeStamp = datum.timeStamp;
       final PointerDeviceKind kind = datum.kind;
       assert(datum.change != null);
-      switch (datum.change) {
-        case ui.PointerChange.add:
-          assert(!_pointers.containsKey(datum.device));
-          final _PointerState state = _ensureStateForPointer(datum, position);
-          assert(state.lastPosition == position);
-          yield PointerAddedEvent(
-            timeStamp: timeStamp,
-            kind: kind,
-            device: datum.device,
-            position: position,
-            obscured: datum.obscured,
-            pressureMin: datum.pressureMin,
-            pressureMax: datum.pressureMax,
-            distance: datum.distance,
-            distanceMax: datum.distanceMax,
-            radiusMin: radiusMin,
-            radiusMax: radiusMax,
-            orientation: datum.orientation,
-            tilt: datum.tilt
-          );
-          break;
-        case ui.PointerChange.hover:
-          final bool alreadyAdded = _pointers.containsKey(datum.device);
-          final _PointerState state = _ensureStateForPointer(datum, position);
-          assert(!state.down);
-          if (!alreadyAdded) {
-            assert(state.lastPosition == position);
-            yield PointerAddedEvent(
-              timeStamp: timeStamp,
-              kind: kind,
-              device: datum.device,
-              position: position,
-              obscured: datum.obscured,
-              pressureMin: datum.pressureMin,
-              pressureMax: datum.pressureMax,
-              distance: datum.distance,
-              distanceMax: datum.distanceMax,
-              radiusMin: radiusMin,
-              radiusMax: radiusMax,
-              orientation: datum.orientation,
-              tilt: datum.tilt
-            );
-          }
+      if (datum.kind == PointerDeviceKind.gesture) {
+        // Devices must be added before they send scroll events.
+        assert(_pointers.containsKey(datum.device));
+        final _PointerState state = _ensureStateForPointer(datum, position);
+        if (state.lastPosition != position) {
+          // Synthesize a hover of the pointer to the scroll location before
+          // sending the scroll event, if necessary, so that clients don't have
+          // to worry about native ordering of hover and scroll events.
           final Offset offset = position - state.lastPosition;
           state.lastPosition = position;
           yield PointerHoverEvent(
```

Comment:
what if the pointer is actually already down? (e.g. scrolling while dragging)

Reply:
Good point, fixed.